src/QsCompiler/Tests.Compiler/LinkingTests.fs

```diff
@@ -11,93 +11,191 @@ open Microsoft.Quantum.QsCompiler.DataTypes
+
+        for signature in targetSignatures do
+            let sig_fullName, sig_argType, sig_rtrnType = signature
+            callableElems
+            |> Seq.exists (fun call -> doesCallMatchSig call signature)
+            |> (fun x -> Assert.True (x, sprintf "Expected but did not find: %s.%s %s : %A" sig_fullName.Namespace.Value sig_fullName.Name.Value (makeArgsString sig_argType) sig_rtrnType.Resolution))
+
+    static member private GetEntryPoints fileName =
+        let validEntryPoints = Path.Combine ("TestCases", "LinkingTests", fileName) |> File.ReadAllText
+        validEntryPoints.Split ([|"==="|], StringSplitOptions.RemoveEmptyEntries)
+
+    member private this.Expect name (diag : IEnumerable<DiagnosticItem>) =
         let ns = "Microsoft.Quantum.Testing.EntryPoints" |> NonNullable<_>.New
         let name = name |> NonNullable<_>.New
         this.Verify (QsQualifiedName.New (ns, name), diag)
 
-    member private this.CompileAndVerify input (diag : DiagnosticItem seq) = 
+    member private this.CompileAndVerify input (diag : DiagnosticItem seq) =
 
         let fileId = getTempFile()
         let file = getManager fileId input
         let inFile (c : QsCallable) = c.SourceFile = file.FileName
-        
-        compilation.AddOrUpdateSourceFileAsync(file) |> ignore
-        let built = compilation.Build()
-        compilation.TryRemoveSourceFileAsync(fileId, false) |> ignore
+
+        compilationManager.AddOrUpdateSourceFileAsync(file) |> ignore
+        let built = compilationManager.Build()
+        compilationManager.TryRemoveSourceFileAsync(fileId, false) |> ignore
         let tests = new CompilerTests(built, output)
 
-        for callable in built.Callables.Values |> Seq.filter inFile do 
+        for callable in built.Callables.Values |> Seq.filter inFile do
             tests.Verify (callable.FullName, diag)
 
+    member private this.CompileAndTestMonomorphization input =
+
+        let fileId = getTempFile()
+        let file = getManager fileId input
+
+        compilationManager.AddOrUpdateSourceFileAsync(file) |> ignore
+        let compilationDataStructures = compilationManager.Build()
+        compilationManager.TryRemoveSourceFileAsync(fileId, false) |> ignore
+
+        compilationDataStructures.Diagnostics() |> Seq.exists (fun d -> d.IsError()) |> Assert.False
+
+        Assert.NotNull compilationDataStructures.BuiltCompilation
+        let mutable monomorphicCompilation = compilationDataStructures.BuiltCompilation
+        compilationDataStructures.BuiltCompilation.Monomorphisize &monomorphicCompilation |> Assert.True
+
+        Assert.NotNull monomorphicCompilation
+        monomorphicCompilation.ValidateMonomorphization () |> Assert.True
+        monomorphicCompilation
+
+
+    [<Fact>]
+    member this.``Monomorphization`` () =
+
+        let makeSig input =
+            let ns, name, args, rtrn = input
+            { Namespace = NonNullable<string>.New ns; Name = NonNullable<string>.New name },
```

Comment:
nit: Hmm, in this case actually assigning the name, argument and return type to variables, and then returning them as `(fullName, argType, returnType)` would make things more readable to me.